tests/codenames/test_codenames_game.py

```diff
@@ -0,0 +1,198 @@
+    elif error == 2: # invalid count
+        count = -10 + random.randint(0, 1) * random.random()
+    elif error == 3: # invalid word and count
+        word = random.choice(state["words"])
+        count = -10 + random.randint(0, 1) * random.random()
+
+    return "invalid", json.dumps({
+        "word": word,
+        "count": count,
+    })
+
+
+def guesser_valid_move(game):
+    state = json.loads(game.get_state()[0])  
+    
+    valid_words = []
+    for idx in range(len(state["words"])):
+        if state["guessed"][idx] != "UNKNOWN" or state["actual"][idx] != "RED":
+            continue
+        valid_words.append(idx)
+    
+    print("valid words: ", valid_words)
+    guess_num = random.randint(1, min(4, len(valid_words)))
+
+    return "valid", json.dumps({
+        "guesses": random.sample(valid_words, guess_num),
+    })
+
+def guesser_invalid_move(game):
+    state = json.loads(game.get_state()[0])  
+
+    valid_words = []
+    non_ours_words = []
+    guessed_words = []
+    assassin_word = None
+
+    for idx in range(len(state["words"])):
+        if state["guessed"][idx] != "UNKNOWN":
+            guessed_words.append(idx)
+
+        if state["actual"][idx] != "RED" and state["guessed"][idx] == "UNKNOWN":
+            non_ours_words.append(idx)
+
+        if state["guessed"][idx] == "UNKNOWN" and state["actual"][idx] == "RED":
+            valid_words.append(idx)
+
+        if state["actual"][idx] == "ASSASSIN":
+            assassin_word = idx
+
+    error = random.randint(1, 5)
+
+    valid_padding = random.sample(valid_words, min(random.randint(1, 3), len(valid_words)))
+
+    # split valid words into two groups
+    valid_words1 = random.sample(valid_padding, random.randint(0, max(0, len(valid_padding) - 1)))
+    valid_words2 = [x for x in valid_padding if x not in valid_words1]
+    
+    if len(guessed_words) == 0 or len(non_ours_words) == 0:
+        error = random.randint(2, 4)
+
+    if error == 1:  # guessed word
+        return "invalid", json.dumps({
+            "guesses": valid_words1 + [random.choice(guessed_words)] + valid_words2,
+        })
+    elif error == 2:  # non-valid word
+        return "invalid", json.dumps({
+            "guesses": valid_words1 + [random.random()] + valid_words2,
+        })
+    elif error == 3:  # non-valid word
+        return "invalid", json.dumps({
+            "guesses": valid_words1 + [-10] + valid_words2,
+        })
+    elif error == 4:  # assassin word
+        
+        print(state)
+        print(assassin_word)
+        print(valid_words1 + [assassin_word] + valid_words2)
+    
+        return "assassin", json.dumps({
+            "guesses": valid_words1 + [assassin_word] + valid_words2,
+        })
+    elif error == 5:  # non-ours word
+        return "not-ours", json.dumps({
+            "guesses": valid_words1 + [random.choice(non_ours_words)] + valid_words2,
+        })
+
+class TestCodenamesDefinition(unittest.TestCase):
+
+    def test_random_game(self):
```

Comment:
actually i think this test is very interesting. We can keep the randomness, but let's also try and make a deterministic test. Maybe we can optionally pass in error to the invalid functions, then we can for loop over the different combinations to make sure we always try each of the error options